### ISSUE TITLE
Update about Keras Embedding Implementation

### ISSUE THREAD

**Issue by jeromemassot** [plaidml/plaidml#91]
Dear developer team,
any update about the implementation of Keras embedding in PlaidML?
Thanks
Best regards
Jerome

**Comment by brianretford** [comment]
Duplicate of #92 
The answer is soon, though. Very soon.

**Comment by brianretford** [comment]
FYI, you can try `pip install plaidml-keras==0.3.0rc1 plaidml==0.3.0rc1` and let us know if it work

**Comment by jeromemassot** [comment]
Hi Brian,
and thank you for your reply and update comment.
I have updated plaidml and now the embedding layer is recognized. The model can compile and be trained. However, there is a problem of performance. The fit operation is very slow, but it appears that the bottleneck could be located at the very end of each epoch. I do not know what Keras is doing near the end of each epoch (may be calculation on validation set) but it slows down the fit process. This behavior has not been observed with NVidia cards.
Now there is another issue with SimpleRNN. When I add a SimpleRNN layer as SimpleRNN(32), I receive a "NotImplementedError" telling that RNN is not implemented for variable sized inputs.
New openticket in approach if not existing yet :)
But first, thanks a lot to the Team for continuing implementing Keras under PlaidML :)
Best regards
Jérôme

**Comment by jbruestle** [comment]
Hi Jérôme.  Thanks for your testing.  There are two issue here.  

The first is the we don't support sparse tensors quite yet, which means that training an embedding layer with lots of embeddings (IE a large vocabulary in the case word embeddings) can be very inefficient.  Support for sparse vectors is something which is on our long term roadmap, but isn't currently scheduled, so I don't know for sure when it will be fixed.

The second is that we currently require the number of time steps to be known at the time that Keras 'compiles' the RNN function.  The workaround for this, which is effective for many but not all RNNs, is to make sure the time dimension of the input shape is an integer rather than a 'None'.  Usually this can be accomplished by setting the input_shape parameter of the input layer to contain the number of time steps in the appropriate dimension.  Hopefully that will work for you.  If not, we will also eventually add support for a dynamic number of time steps in recurrent nets, but again, I don't have a specific timeline.

As for whether to open two separate issues for this issue, I'll leave that up to Brian, although my inclination work be to do so.

-Jeremy

**Comment by brianretford** [comment]
We've created alternate issues that capture the remaining work here.
